drv/lpc55-update-server/src/main.rs

```diff
@@ -402,6 +387,132 @@ impl idl::InOrderUpdateImpl for ServerImpl<'_> {
     }
 }
 
+impl ServerImpl<'_> {
+    // Locate and return the authoritative CFPA flash word number
+    // and the CFPA version for that flash number.
+    //
+    // There are two "official" copies of the CFPA, referred to as
+    // ping and pong. One of them will supercede the other, based on
+    // a monotonic version field at offset 4. We'll take the
+    // contents of whichever one is most recent, alter them, and
+    // then write them into the _third_ copy, called the scratch
+    // page.
+    //
+    // At reset, the boot ROM will inspect the scratch page, check
+    // invariants, and copy it to overwrite the older of the ping
+    // and pong pages if it approves.
+    //
+    // That means you can apply this operation several times before
+    // resetting without burning many monotonic versions, if you
+    // want to do that for some reason.
+    //
+    // The addresses of these pages are as follows (see Figure 13,
+    // "Protected Flash Region," in UM11126 rev 2.4, or the NXP
+    // flash layout spreadsheet):
+    //
+    // Page     Addr        16-byte word number
+    // Scratch  0x9_DE00    0x9DE0
+    // Ping     0x9_E000    0x9E00
+    // Pong     0x9_E200    0x9E20
+    fn cfpa_word_number_and_version(
+        &mut self,
+    ) -> Result<(u32, u32), UpdateError> {
+        // Read the two versions. We do this with smaller buffers so
+        // we don't need 2x 512B buffers to read the entire CFPAs.
+        let mut ping_header = [0u32; 4];
+        let mut pong_header = [0u32; 4];
+
+        indirect_flash_read_words(
+            &mut self.flash,
+            CFPA_PING_FLASH_WORD,
+            core::slice::from_mut(&mut ping_header),
+        )?;
+        indirect_flash_read_words(
+            &mut self.flash,
+            CFPA_PONG_FLASH_WORD,
+            core::slice::from_mut(&mut pong_header),
+        )?;
+
+        // Work out where to read the authoritative contents from.
+        let val = if ping_header[1] >= pong_header[1] {
+            (CFPA_PING_FLASH_WORD, ping_header[1])
+        } else {
+            (CFPA_PONG_FLASH_WORD, pong_header[1])
+        };
+
+        Ok(val)
+    }
+
+    // Return the persistent and transient boot preferences
+    fn boot_preferences(
+        &mut self,
+    ) -> Result<(SlotId, Option<SlotId>, Option<SlotId>), UpdateError> {
+        let (cfpa_word_number, cfpa_version) =
+            self.cfpa_word_number_and_version()?;
+
+        // Read the authoritative boot selection
+        let boot_selection_word_number = cfpa_word_number + 0x10;
```

Comment:
Add a comment (or a `const`, since it's used below as well) explaining where this offset comes from

Reply:
`const` added in [3c3ca2e](https://github.com/oxidecomputer/hubris/pull/1367/commits/3c3ca2ea4042426ff64a3f7f6f947fb5f910f10c)